File: Sorting/heapsort.py
"""
Copyright (c) 2017 Oscar Albornoz. All rights reserved.
"""

import random
import time

def heap_sort(arr):
    max_heapify(arr)
    for end in reversed(range(len(arr))):
        arr[end], arr[0] = arr[0], arr[end]
        sift_down(arr,0,end-1)
    return arr

def max_heapify(arr):
    mid = (len(arr) >> 1)
    for i in reversed(range(mid)):
        sift_down(arr,i,len(arr)-1)

def sift_down(arr, father, end):
    while True:
        left_child = (father << 1) + 1
        right_child = left_child + 1

        if left_child > end:
            break

        if right_child > end:
            right_child = left_child

        if arr[left_child] >= arr[right_child]:
            greater_child = left_child
        else:
            greater_child = right_child

        if arr[father] >= arr[greater_child]:
                break

        arr[greater_child], arr[father] = arr[father], arr[greater_child]
        father = greater_child

def main():
    """
    Testing the function
    """
    random.seed(1)
    arr = list(random.randint(-5000, 100000) for i in range(100000))
    copy_one = copy_two = arr

    start = time.clock()
    copy_one = heap_sort(arr)
    print(time.clock() - start)

    start = time.clock()
    copy_two = sorted(copy_two)
    print(time.clock()-start)

    print(copy_one == copy_two)

if __name__ == "__main__":
    main()

File: Sorting/heap_sort.cc
// Copyright (c) 2017 Oscar Albornoz. All rights reserved.


#include <stdlib.h>
#include <time.h>
#include <ctime>
#include <iostream>

template <typename T, typename Comp = std::less<T>>
void heap_sort(T* array, int size, const Comp &cmp = Comp());

template <typename T, typename Comp = std::less<T>>
void heapify(T* array, int size, const Comp &cmp = Comp());

template <typename T, typename Comp = std::less<T>>
void sift_down(T* array, int parent, int end, const Comp &cmp = Comp());

using std::cout;
using std::endl;

const int kSize = 1000000;

int main(int argc, char const *argv[]) {
  int arr[kSize];

  for (int i = 0; i < kSize; ++i) {
    arr[i] = rand() % kSize;  // NOLINT
  }

  cout << "Sorting " << kSize << " elements" << endl;

  clock_t begin = clock();

  heap_sort(arr, kSize, [](int f, int s) {return f > s;});

  clock_t end = clock();
  double elapsed_secs = static_cast<double>(end - begin) / CLOCKS_PER_SEC;

  cout << "Time: " << elapsed_secs << endl;

  return 0;
}

template <typename T, typename Comp>
void heap_sort(T* array, int size, const Comp &cmp) {
  heapify(array, size, cmp);

  for (int end = size - 1; end >= 0; --end) {
    std::swap(array[0], array[end]);
    sift_down(array, 0, end - 1, cmp);
  }
}

template <typename T, typename Comp>
void heapify(T* array, int size, const Comp &cmp) {
  for (int i = size / 2 - 1; i >= 0; --i) {
    sift_down(array, i, size - 1, cmp);
  }
}

template <typename T, typename Comp>
void sift_down(T* array, int parent, int end, const Comp &cmp) {
  int left_child, right_child, swappable_child;

  while ((left_child = (parent << 1) + 1) <= end) {
    right_child = left_child + 1;

    if (right_child > end)
      right_child = left_child;

    swappable_child = cmp(array[right_child], array[left_child]) ?
      left_child : right_child;

    if (cmp(array[swappable_child], array[parent]))
      break;

    std::swap(array[parent], array[swappable_child]);
    parent = swappable_child;
  }
}
